### PR TITLE
feat: entity store

### LINKED ISSUE #130: Add dependency injection support
Add support for dependency injection. One consideration is being able to register an `EventMapper` in the container so it can be retrieved for read-side concerns.

### LINKED ISSUE #131: Add initial code to develop-v1 branch

### LINKED ISSUE #132: Introduce concurrency exception in library
Right now, we're throwing the underlying infrastructure's concurrency exceptions. DomainBlocks should wrap underlying infrastructure exceptions like this to enable consistent exception handling for consumers.

### PR DESCRIPTION
**Pull request by daniel-smith**
Initial PR to `develop-v1` branch.

Introduces a new `EntityStore`, replacing the existing `AggregateRepository`. Includes a flexible approach to "adapt" any type of event sourced entity to a given EntityStore instance.

This is only an initial PR. More changes to come as part of the [Version 1 project](https://github.com/orgs/DomainBlocks/projects/5).

Closes #130, closes #131, closes #132

### CODE REVIEW

File: examples/shopping/Shopping.ReadModel/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                        .ConfigureSqlStreamStoreProjection(sp, connectionString)
+                        .ConfigureEventHandlers();
+                    break;
+            }
+        });
+
+        return services;
+    }
+
+    private static StateProjectionOptionsBuilder<ResolvedEvent, Position, ShoppingCartDbContext>
```
Comment (daniel-smith) [inline examples/shopping/Shopping.ReadModel/ServiceCollectionExtensions.cs:52]: FYI, this is all quite messy due to how the read side builders are currently designed (and how the read side is designed in general). Will be addressed in #136.

File: src/DomainBlocks.Persistence.EventStoreDb/EventStoreDbEventStore.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        try
+        {
+            var eventStoreDbDirection =
+                direction == StreamReadDirection.Forwards ? Direction.Forwards : Direction.Backwards;
+
+            var eventStoreDbFromVersion = fromVersion != null
+                ? new StreamPosition(fromVersion.Value.ToUInt64())
+                : readOrigin!.Value.ToEventStoreDbStreamPosition(direction);
+
+            // Consider reading in batches.
```
Comment (daniel-smith) [inline src/DomainBlocks.Persistence.EventStoreDb/EventStoreDbEventStore.cs:76]: I don't think this comment is relevant anymore, given the EventStoreDB client returns `IAsyncEnumerable`.

**Review by jonclare — DISMISSED** [review]
Few minor comments

File: src/DomainBlocks.Persistence/Entities/EntityAdapterRegistry.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            adapter = (IEntityAdapter<TEntity>)result;
+            return true;
+        }
+
+        adapter = _genericAdapterFactories
+            .Select(x =>
+            {
+                var success = x.TryCreateFor<TEntity>(out var instance);
+                return (success, instance);
```
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/Entities/EntityAdapterRegistry.cs:37]: I think the code style on this is a bit divided. On the one hand, these can still be considered "local", and there's a school of thought that they should be `camelCase` in that case. I'm not particularly partial to one way or another, but if we rename them to Pascal Case, we'd have to do `return (Success: success, Instance: instance)`, which arguably adds unnecessary noise. If these are returned from a method, then for sure Pascal Case. Just not sure about local usage like this. One to discuss!

File: src/DomainBlocks.EventStore/EventStoreServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -26,7 +26,7 @@ public static IServiceCollection AddEventStore(this IServiceCollection services,
                     {
                         var options = provider.GetRequiredService<IOptions<EventStoreConnectionOptions>>();
 
-                        var settings = EventStoreClientSettings.Create(options.Value.ConnectionString);
+                        var settings = EventStoreClientSettings.Create(options.Value.ConnectionString!);
```
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.EventStore/EventStoreServiceCollectionExtensions.cs:29]: Quite possibly. Will have a look why this is here.

File: src/DomainBlocks.Persistence/EntityStore.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        // Used in closure of TransformEventStream, so must be declared before the async enumerable is materialized,
+        // i.e. before RestoreEntity is invoked.
+        StreamVersion? loadedVersion = null;
+        var loadedEventCount = 0;
+
+        var entity = await entityAdapter.RestoreEntityAsync(initialState, TransformEventStream(), cancellationToken);
+        var trackedEntityContext = new TrackedEntityContext(loadedVersion, loadedEventCount);
+
+        // Track the entity so that the expected version will be known in a future call to SaveAsync.
+        _trackedEntities.Add(entity!, trackedEntityContext);
```
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/EntityStore.cs:44]: It seems to be due to the fact that the key for `ConditionalWeakTable` is `object`. `RestoreEntityAsync` is guaranteed to return something non-null, at least according to nullable reference types. I actually don't fully understand the details around why it complains in this particular case.

[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/EntityStore.cs:44]: We can throw or assert to get rid of the null forgiving operator in this case.

File: src/DomainBlocks.Persistence/Entities/EntityAdapterRegistry.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonclare) [reply src/DomainBlocks.Persistence/Entities/EntityAdapterRegistry.cs:37]: I prefer doing `return (Success: success, Instance: instance)`
If you're not going to reference the fields of the tuple anywhere, then I would agree it's fine.
If you are, I'd prefer them to be Pascal Case.
Either way, we should be consistent whatever we decide here. Would be interested to see if there's a well-known preferred style for these things

[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/Entities/EntityAdapterRegistry.cs:37]: From what I've seen, the opinions are fairly divided on this. Both sides have valid points. But yes, agree that we need to be consistent. That leads to the wider topic of having a defined coding style on this project.

[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/Entities/EntityAdapterRegistry.cs:37]: To "properly" solve this issue, we'd need to add the `where TEntity : class` constraint everywhere. Not entirely sure we want this thought. I "think" it is a reasonable assumption that all entities will be represented by reference types, but right now we don't have this constraint.

File: src/DomainBlocks.Persistence/EntityStore.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/EntityStore.cs:44]: To "properly" solve this issue, we'd need to add the `where TEntity : class` constraint everywhere. Not entirely sure we want this thought. I "think" it is a reasonable assumption that all entities will be represented by reference types, but right now we don't have this constraint.

File: src/DomainBlocks.Persistence/Extensions/TypeExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        var sb = new StringBuilder();
+        sb.Append(type.Name[..type.Name.IndexOf('`')]);
+        sb.Append('<');
+        sb.Append(string.Join(", ", type.GetGenericArguments().Select(GetPrettyName)));
+        sb.Append('>');
+
+        return sb.ToString();
+    }
+
+    public static bool HasInterface(this Type type, Type interfaceType)
```
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/Extensions/TypeExtensions.cs:20]: Although, to be pedantic, an interface can inherit another interface. In this case, it's not implemented.

**Review by jonclare — APPROVED** [review]

File: src/DomainBlocks.Persistence/EntityStore.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Persistence/EntityStore.cs]: As discussed, using the `notnull` constraint on `TEntity` and `TState`.

**Pull request merged**